### ISSUE TITLE
Crash in flecs when dealing with a lot of entities and using System.OrderBy

### ISSUE THREAD

**Issue by deathbeam** [BeanCheeseBurrito/Flecs.NET#45]
lldb backtrace:

```
(lldb) bt
* thread #1, name = 'Game', stop reason = signal SIGSEGV: address not mapped to object
  * frame #0: 0x0000757b768709f5 libcoreclr.so`___lldb_unnamed_symbol9797 + 9749
    frame #1: 0x0000757b768696a0 libcoreclr.so`___lldb_unnamed_symbol9791 + 1024
    frame #2: 0x0000757b76876c3b libcoreclr.so`___lldb_unnamed_symbol9834 + 1723
    frame #3: 0x0000757b76864699 libcoreclr.so`___lldb_unnamed_symbol9727 + 1033
    frame #4: 0x0000757b76865cba libcoreclr.so`___lldb_unnamed_symbol9739 + 58
    frame #5: 0x0000757b76866919 libcoreclr.so`___lldb_unnamed_symbol9742 + 601
    frame #6: 0x0000757b76897860 libcoreclr.so`___lldb_unnamed_symbol10069 + 208
    frame #7: 0x0000757b7671ac7f libcoreclr.so`___lldb_unnamed_symbol5652 + 175
    frame #8: 0x0000757b7671be03 libcoreclr.so`___lldb_unnamed_symbol5663 + 115
    frame #9: 0x0000757b76738166 libcoreclr.so`___lldb_unnamed_symbol6131 + 198
    frame #10: 0x0000757af8b09324
    frame #11: 0x0000757af8b08fce
    frame #12: 0x0000757af8b08f0d
    frame #13: 0x0000757b74f7cf71 libflecs.so`flecs_run_intern(world=0x00005d5c2b7e40b0, stage=0x00005d5c2b7fb150, system=104721, system_data=0x00005d5c2e227520, stage_index=0, stage_count=1, delta_time=0.230314463, param=0x0000000000000000) at flecs.c:63446:21
    frame #14: 0x0000757b74f7cabf libflecs.so`flecs_run_pipeline_ops(world=0x00005d5c2b7e40b0, stage=0x00005d5c2b7fb150, stage_index=0, stage_count=1, delta_time=0.230314463) at flecs.c:52930:9
    frame #15: 0x0000757b74f7d5e7 libflecs.so`flecs_run_pipeline(world=0x00005d5c2b7e40b0, pq=0x00005d5c2b9715b0, delta_time=0.230314463) at flecs.c:52999:27
    frame #16: 0x0000757b74f7c43f libflecs.so`flecs_workers_progress(world=0x00005d5c2b7e40b0, pq=0x00005d5c2b9715b0, delta_time=0.230314463) at flecs.c:53542:5
    frame #17: 0x0000757b74f7da6f libflecs.so`ecs_progress(world=0x00005d5c2b7e40b0, user_delta_time=0.230314463) at flecs.c:53115:5
    frame #18: 0x0000757af8b03070
    frame #19: 0x0000757af8b02f7d
    frame #20: 0x0000757af8221f10
    frame #21: 0x0000757b7689fcc7 libcoreclr.so`___lldb_unnamed_symbol10269 + 124
    frame #22: 0x0000757b766d833e libcoreclr.so`___lldb_unnamed_symbol4841 + 1694
    frame #23: 0x0000757b765beb44 libcoreclr.so`___lldb_unnamed_symbol926 + 820
    frame #24: 0x0000757b765befb8 libcoreclr.so`___lldb_unnamed_symbol928 + 456
    frame #25: 0x0000757b765ec00c libcoreclr.so`___lldb_unnamed_symbol1627 + 732
    frame #26: 0x0000757b765a9380 libcoreclr.so`coreclr_execute_assembly + 144
    frame #27: 0x0000757b770cba6e libhostpolicy.so`___lldb_unnamed_symbol537 + 910
    frame #28: 0x0000757b770ccb89 libhostpolicy.so`corehost_main + 345
    frame #29: 0x0000757b7710a333 libhostfxr.so`___lldb_unnamed_symbol359 + 1523
    frame #30: 0x0000757b7710930d libhostfxr.so`___lldb_unnamed_symbol357 + 765
    frame #31: 0x0000757b771055a2 libhostfxr.so`hostfxr_main_startupinfo + 242
    frame #32: 0x00005d5c18bfbb10 Game`___lldb_unnamed_symbol354 + 1136
    frame #33: 0x00005d5c18bfbe3f Game`___lldb_unnamed_symbol356 + 175
    frame #34: 0x0000757b76c34e08 libc.so.6`___lldb_unnamed_symbol3261 + 120
    frame #35: 0x0000757b76c34ecc libc.so.6`__libc_start_main + 140
    frame #36: 0x00005d5c18bf20f9 Game`___lldb_unnamed_symbol213 + 41
```

Not sure what info is needed, but this happens in systems, i used same code (mostly) with other ECS (https://github.com/andreakarasho/TinyEcs for example, also tried with Arch) just with different way of sorting.

If I remove the OrderBy i prevent the crashing. If i leave the OrderBy body empty and always return 0 the crash still happens so the issue isnt with my code inside. On similar note. I also get crashes on much lower number of entities when using Multithreaded.

The entity count is 100k. I also get crash when bumping the amount of entities to 150k even without .OrderBy but that might be different issue.

Here are the components:

```csharp
using System.Numerics;

public record struct Position(float X, float Y, float Z = 0)
{
    public static Position operator +(Position a, Translate b) => new Position(a.X + b.X, a.Y + b.Y);
}

public record struct Translate(float X, float Y);

public record struct Scale(float X, float Y); // Unused for now

public record struct Velocity(float X, float Y);

public record struct Source(int X, int Y, int Width, int Height);

public record struct Texture(string Path);

public record struct Light(float Radius = 256);

public record struct Occluder(Vector2[] Points);

public record struct Collider(Vector2[] Points);

public record struct GridPath(Vector2[] Points, int Progress);

public record struct Destination(float X, float Y);

public record struct Camera(float TargetX, float TargetY, float OffsetX, float OffsetY, float Zoom)
{
    public (float Left, float Top, float Right, float Bottom) GetBounds(float offset)
    {
        float left = TargetX - (OffsetX / Zoom) - offset;
        float right = TargetX + (OffsetX / Zoom) + offset;
        float top = TargetY - (OffsetY / Zoom) - offset;
        float bottom = TargetY + (OffsetY / Zoom) + offset;
        return (left, top, right, bottom);
    }
}
```

And here are the systems:

```csharp
using System.Numerics;
using Flecs.NET.Core;
using Raylib_cs;

public class Systems
{
    public static void UpdateCamera(ref World world, Renderer renderer)
    {
        void Run(Iter it, int i, ref Camera camera)
        {
            var delta = it.DeltaTime();
            var moveSpeed = 500 * delta;
            var zoomSpeed = 5 * delta;

            var size = renderer.WindowSize();
            camera.OffsetX = size.X / 2;
            camera.OffsetY = size.Y / 2;

            if (Raylib.IsKeyDown(KeyboardKey.W))
                camera.TargetY -= moveSpeed;
            if (Raylib.IsKeyDown(KeyboardKey.S))
                camera.TargetY += moveSpeed;
            if (Raylib.IsKeyDown(KeyboardKey.A))
                camera.TargetX -= moveSpeed;
            if (Raylib.IsKeyDown(KeyboardKey.D))
                camera.TargetX += moveSpeed;

            if (Raylib.IsKeyDown(KeyboardKey.KpAdd) || Raylib.IsKeyDown(KeyboardKey.Equal))
                camera.Zoom += zoomSpeed;
            if (Raylib.IsKeyDown(KeyboardKey.KpSubtract) || Raylib.IsKeyDown(KeyboardKey.Minus))
                camera.Zoom -= zoomSpeed;

            renderer.UpdateCamera(camera);
        }

        world.System<Camera>("UpdateCamera").TermAt(0).Singleton().Each(Run);
    }

    public static void DisableEntities(ref World world, Grid grid)
    {
        void Run(Iter it, Field<Position> positions)
        {
            var camera = it.World().Get<Camera>();
            var offset = grid.TileSize * 2; // TODO: Account for texture size?
            var (left, top, right, bottom) = camera.GetBounds(offset);

            foreach (int i in it)
            {
                var pos = positions[i];
                var entity = it.Entity(i);
                if (!entity.IsValid())
                {
                    continue;
                }

                if (pos.X < left || pos.X > right || pos.Y < top || pos.Y > bottom)
                {
                    entity.Disable();
                }
                else
                {
                    entity.Enable();
                }
            }
        }

        world.System<Position>("DisableEntities").With(Ecs.Disabled).Optional().Iter(Run);
    }

    public static void Move(ref World world)
    {
        void Run(Iter it, Field<Velocity> velocities, Field<Position> positions)
        {
            foreach (int i in it)
            {
                positions[i].X += it.DeltaTime() * velocities[i].X;
                positions[i].Y += it.DeltaTime() * velocities[i].Y;
            }
        }

        world.System<Velocity, Position>("Move").Iter(Run);
    }

    public static void RandomDestination(ref World world, Random random)
    {
        void Run(Iter it, Field<Position> positions, Field<Destination> destinations, Field<Velocity> velocities)
        {
            foreach (int i in it)
            {
                velocities[i].X = 0;
                velocities[i].Y = 0;
                destinations[i].X = positions[i].X + random.Next(-100, 100);
                destinations[i].Y = positions[i].Y + random.Next(-100, 100);
            }
        }

        world.System<Position, Destination, Velocity>("RandomDestination").Without<GridPath>().Iter(Run);
    }

    public static void FollowPaths(ref World world, Grid grid)
    {
        void Run(Iter it, Field<Position> positions, Field<GridPath> paths, Field<Velocity> velocities)
        {
            foreach (int i in it)
            {
                var pos = positions[i];
                var vel = velocities[i];
                var path = paths[i];
                var entity = it.Entity(i);
                if (!entity.IsValid())
                {
                    continue;
                }

                var gridPos = grid.GetNode(pos.X, pos.Y);
                if (gridPos == null)
                {
                    continue;
                }

                if (path.Progress < path.Points.Length)
                {
                    var target = path.Points[path.Progress];
                    var direction = new Vector2(target.X - gridPos.X, target.Y - gridPos.Y);
                    var distance = direction.Length();
                    direction /= distance; // Normalize the direction vector

                    if (target == gridPos.Position)
                    {
                        paths[i].Progress++;
                    }
                    else
                    {
                        velocities[i].X = direction.X * 100;
                        velocities[i].Y = direction.Y * 100;
                    }
                }
                else
                {
                    entity.Remove<GridPath>();
                }
            }
        }

        world.System<Position, GridPath, Velocity>("FollowPaths").Iter(Run);
    }

    public static void FindPaths(ref World world, Grid grid)
    {
        var tasks = new Tasks<ulong, Vector2[]?>();

        void Run(Iter it, Field<Position> positions, Field<Destination> destinations)
        {
            foreach (int i in it)
            {
                var pos = positions[i];
                var target = destinations[i];
                var startNode = grid.GetNode(pos.X, pos.Y);
                var endNode = grid.GetNode(target.X, target.Y);

                if (startNode != null && endNode != null)
                {
                    var entity = it.Entity(i);
                    if (!entity.IsValid())
                    {
                        continue;
                    }

                    var path = tasks.Run(entity.Id, () => grid.FindPath(startNode, endNode));
                    if (path != null)
                    {
                        entity.Set<GridPath>(new GridPath(path, 1));
                    }
                }
            }
        }

        world.System<Position, Destination>("FindPaths").Iter(Run);
    }

    public static void ProcessOccluders(ref World world, Renderer renderer)
    {
        void Run(Iter it, Field<Position> positions, Field<Occluder> occluders, Field<Translate> translates)
        {
            foreach (int i in it)
            {
                var occ = occluders[i];
                var pos = positions[i];
                if (it.IsSet(2))
                {
                    pos += translates[i];
                }

                renderer.AddOccluder(pos, occ);
            }
        }

        world.System<Position, Occluder, Translate>("ProcessOccluders").TermAt(2).Optional().Iter(Run);
    }

    public static void ProcessColliders(ref World world, Grid grid)
    {
        void Run(Iter it, Field<Position> positions, Field<Collider> colliders, Field<Translate> translates)
        {
            foreach (int i in it)
            {
                var entity = it.Entity(i);
                if (!entity.IsValid())
                {
                    continue;
                }

                var id = entity.Id.Value;
                var pos = positions[i];
                var col = colliders[i];
                if (it.IsSet(2))
                {
                    pos += translates[i];
                }
                grid.UpdateEntityPosition(id, pos.X, pos.Y, col.Points);
            }
        }

        world.System<Position, Collider, Translate>("ProcessColliders").TermAt(2).Optional().Iter(Run);
    }

    public static System<Light, Position, Color, Translate> DrawLights(ref World world, Renderer renderer)
    {
        void Run(
            Iter it,
            Field<Light> lights,
            Field<Position> positions,
            Field<Color> colors,
            Field<Translate> translates
        )
        {
            foreach (int i in it)
            {
                var pos = positions[i];
                if (it.IsSet(3))
                {
                    pos += translates[i];
                }

                renderer.DrawLight(lights[i], pos, colors[i]);
            }
        }

        return world.System<Light, Position, Color, Translate>("DrawLights").TermAt(3).Optional().Kind(0).Iter(Run);
    }

    public static unsafe System<Texture, Position, Source, Color, Translate> DrawTextures(
        ref World world,
        Renderer renderer
    )
    {
        int ComparePositions(ulong e1, void* p1, ulong e2, void* p2)
        {
            if (p1 == null || p2 == null || e1 == 0 || e2 == 0)
            {
                return 0;
            }

            var pos1 = (Position*)p1;
            var pos2 = (Position*)p2;
            int zComparison = pos1->Z.CompareTo(pos2->Z);
            if (zComparison != 0)
            {
                return zComparison;
            }
            return e1.CompareTo(e2);
        }

        void Run(
            Iter it,
            Field<Texture> textures,
            Field<Position> positions,
            Field<Source> sources,
            Field<Color> colors,
            Field<Translate> translates
        )
        {
            foreach (int i in it)
            {
                var pos = positions[i];
                if (it.IsSet(4))
                {
                    pos += translates[i];
                }

                renderer.DrawTexture(textures[i], pos, sources[i], colors[i]);
            }
        }

        return world
            .System<Texture, Position, Source, Color, Translate>("DrawTextures")
            .TermAt(4)
            .Optional()
            .Kind(0)
            .OrderBy<Position>(ComparePositions)
            .Iter(Run);
    }

    public static System<GridPath> DrawPaths(ref World world, Grid grid)
    {
        void Run(Iter it, Field<GridPath> paths)
        {
            foreach (int i in it)
            {
                var path = paths[i].Points;

                for (int j = 0; j < path.Length - 1; j++)
                {
                    var a = path[j];
                    var b = path[j + 1];
                    Raylib.DrawLineEx(grid.GetPosition(a.X, a.Y), grid.GetPosition(b.X, b.Y), 2, Color.SkyBlue);
                }
            }
        }

        return world.System<GridPath>("DrawPaths").Kind(0).Iter(Run);
    }

    public static System_ DrawGrid(ref World world, Grid grid)
    {
        void Run(Iter it)
        {
            for (int x = 0; x < grid.Width; x++)
            {
                for (int y = 0; y < grid.Height; y++)
                {
                    var node = grid.GetNode(x * grid.TileSize, y * grid.TileSize);
                    if (node != null)
                    {
                        Raylib.DrawRectangleLines(
                            node.X * grid.TileSize,
                            node.Y * grid.TileSize,
                            grid.TileSize,
                            grid.TileSize,
                            node.Traverseable ? Color.RayWhite : Color.Red
                        );
                    }
                }
            }
        }

        return world.System("DrawGrid").Kind(0).Iter(Run);
    }
}
```

**Comment by BeanCheeseBurrito** [comment]
There is currently a problem with native callbacks caused by Flecs.NET passing managed function pointers to flecs instead of unmanaged function pointers. I think your crash might be related to this issue. It may take a while to fix.

**Comment by deathbeam** [comment]
> There is currently a problem with native callbacks caused by Flecs.NET passing managed function pointers to flecs instead of unmanaged function pointers. I think your crash might be related to this issue. It may take a while to fix.

And i assume only unamanged being available is tracked in #28? I would try to contribute implementation for it but I was going through the source and its really hard to tell where should I even look. I found Delegates.cs and QueryBuilder.cs in generators but there is bunch of glue code elsewhere too (I was looking at GroupBy-related stuff as reference).

**Comment by deathbeam** [comment]
After finding the root cause for #49 I retested this as well and can confirm that its not happening anymore and its stable. So I guess the actual issue is the managed types in components and not the function delegate?

**Comment by deathbeam** [comment]
Fixed by #54, closing
